### ISSUE TITLE
iot: Unable to list all devices in a registry

### ISSUE THREAD

**Issue by rajasd27** [googleapis/google-cloud-go#7263]
**Client**

IoTCore

**Environment**

MacOS

**Go Environment**

$ go version -  ``` go1.18.3 darwin/amd64 ```

**Code**

e.g.
```go
package main

func fetchAllDevices(ctx context.Context, client *gcpiotcore.DeviceManagerClient, parent string) ([]*gcpiotpb.Device) {

         var fields = &fieldmaskpb.FieldMask{
	        Paths: []string{
		        "id",
		        "name",
		        "credentials",
		        "last_heartbeat_time",
		        "last_event_time",
		        "last_state_time",
		        "last_config_ack_time",
		        "last_config_send_time",
		        "blocked",
		        "last_error_time",
		        "last_error_status",
		        "config",
		        "state",
		        "log_level",
		        "metadata",
		        "gateway_config"},
        }

	req := &gcpiotpb.ListDevicesRequest{
		Parent:       parent,
		FieldMask:    fields,
	}

	var devices []*gcpiotpb.Device

	it := client.ListDevices(ctx, req)

	for {
		resp, err := it.Next()
		if err == iterator.Done {
			break
		}

		if err != nil {
			log.Fatalln("Unable to iterate over device records: ", err)
		}

		devices = append(devices, resp)
	}

	defer client.Close()
	fmt.Println(string(colorGreen), "\u2713 Fetched", len(devices), "devices!", string(colorReset))
	return devices
}
```

**Expected behavior**

```ListDevices``` should return all the devices in the registry (**22766** devices in my case). 

**Actual behavior**

```ListDevices``` returns only **10966** out of 22766 devices. I have another registry where I have more than 200k devices, but the function only returns 19,129 devices.

**Additional context**

Tried logging response returned by  **it.Next()** to see if it has a *NextPageToken* value, but at the end of all the iterations, it always returns nil.

**Comment by quartzmo** [comment]
@rajasd27 Which version of the `cloud.google.com/go/iot/apiv1` client are you using? Did `ListDevices` work correctly for you in the past, with this version or a previous version?

**Comment by rajasd27** [comment]
Hi, thank you for your response. The version that I am using is v1.2.0. I tried upgrading to the latest (v1.5.0) version but still found the same issue.

**Comment by rajasd27** [comment]
No it did not.

**Comment by quartzmo** [comment]
Also, can you update your example to show how you construct the `parent` value, thanks.

**Comment by rajasd27** [comment]
We started seeing this issue a couple of weeks back. 

This is how the parent is being constructed - 
`parent := "projects/" + getProjectID(val) + "/locations/" + Args.gcpRegistryRegion + "/registries/" + Args.registryName`
 
`getProjectID(val)` -> function that extracts project ID from the gcp service account file
`Args.gcpRegistryRegion & Args.registryName` -> cli flags

**Comment by quartzmo** [comment]
I attempted to reproduce this issue with 22000 devices as follows. I could not reproduce the error. Please let me know if you think I can improve the script in any way.

```go
package main

import (
	"context"
	"fmt"

	iot "cloud.google.com/go/iot/apiv1"
	iotpb "cloud.google.com/go/iot/apiv1/iotpb"
	"google.golang.org/api/iterator"
)

func main() {
	ctx := context.Background()
	c, err := iot.NewDeviceManagerClient(ctx)
	if err != nil {
		panic(err)
	}
	defer c.Close()

	project := "..."
	registry := "golang-test-registry-dffabc95-2800-410b-8aa1-1d0a99747762"
	parent := fmt.Sprintf("projects/%s/locations/us-central1/registries/%s", project, registry)
	// for i := 1; i < 22001; i++ {
	// 	device := &iotpb.Device{
	// 		Id: fmt.Sprintf("golang-test-device-%d", i),
	// 	}
	// 	req := &iotpb.CreateDeviceRequest{
	// 		Parent: parent,
	// 		Device: device,
	// 	}
	// 	resp, err := c.CreateDevice(ctx, req)
	// 	if err != nil {
	// 		panic(err)
	// 	}
	// 	fmt.Println("resp: ", resp)
	// 	time.Sleep(50 * time.Millisecond)
	// }

	for i := 1; i < 11; i++ {
		devices := make(map[string]*iotpb.Device)

		reqList := &iotpb.ListDevicesRequest{
			Parent: parent,
		}
		it := c.ListDevices(ctx, reqList)
		for {
			d, err := it.Next()
			if err == iterator.Done {
				fmt.Print("iterator.Done ")
				break
			}
			if err != nil {
				panic(err)
			}
			devices[d.Id] = d
		}
		fmt.Printf("attempt %d total: %d\n", i, len(devices))
	}
}
```

The output:

```go
resp:  id:"golang-test-device-21999"  name:"projects/.../locations/us-central1/registries/golang-test-registry-dffabc95-2800-410b-8aa1-1d0a99747762/devices/2724780632432557"  num_id:2724780632432557  config:{version:1  cloud_update_time:{seconds:1674596008  nanos:819953000}}
resp:  id:"golang-test-device-22000"  name:"projects/.../locations/us-central1/registries/golang-test-registry-dffabc95-2800-410b-8aa1-1d0a99747762/devices/2723405160824226"  num_id:2723405160824226  config:{version:1  cloud_update_time:{seconds:1674596008  nanos:988537000}}

➜  iot go run main.go
iterator.Done attempt 1 total: 22000
iterator.Done attempt 2 total: 22000
iterator.Done attempt 3 total: 22000
iterator.Done attempt 4 total: 22000
iterator.Done attempt 5 total: 22000
iterator.Done attempt 6 total: 22000
iterator.Done attempt 7 total: 22000
iterator.Done attempt 8 total: 22000
iterator.Done attempt 9 total: 22000
iterator.Done attempt 10 total: 22000
```

**Comment by rajasd27** [comment]
I tried using your code snippet and replaced the project & registry details with mine and I'm still seeing the same issue. I tried testing with v1.0.5. What version are you using to run this snippet?

**Comment by rajasd27** [comment]
This is what my code looks like - 

```
package main

import (
	gcpiotcore "cloud.google.com/go/iot/apiv1"
	iotpb "cloud.google.com/go/iot/apiv1/iotpb"
	"context"
	"fmt"
	"google.golang.org/api/iterator"
	"google.golang.org/api/option"
	"log"
)

func main() {
	ctx := context.Background()
	c, err := gcpiotcore.NewDeviceManagerClient(ctx, option.WithCredentialsFile("/Users/user/Downloads/service.json"))
	if err != nil {
		log.Fatalln(err)
	}

	project := "..."
	registry := "iot-core-migration-test"
	parent := fmt.Sprintf("projects/%s/locations/us-central1/registries/%s", project, registry)

	for i := 1; i < 11; i++ {
		devices := make(map[string]*iotpb.Device)

		reqList := &iotpb.ListDevicesRequest{
			Parent: parent,
		}
		it := c.ListDevices(ctx, reqList)
		for {
			d, err := it.Next()
			if err == iterator.Done {
				fmt.Print("iterator.Done ")
				break
			}
			if err != nil {
				panic(err)
			}
			devices[d.Id] = d
		}
		fmt.Printf("attempt %d total: %d\n", i, len(devices))
	}
}
```

**Comment by quartzmo** [comment]
Here's my `go.mod`:

```
module example.com/iot

go 1.19

require cloud.google.com/go/iot v1.5.0

require (
	cloud.google.com/go/compute v1.15.1 // indirect
	cloud.google.com/go/compute/metadata v0.2.3 // indirect
	github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da // indirect
	github.com/golang/protobuf v1.5.2 // indirect
	github.com/google/go-cmp v0.5.9 // indirect
	github.com/googleapis/enterprise-certificate-proxy v0.2.1 // indirect
	github.com/googleapis/gax-go/v2 v2.7.0 // indirect
	go.opencensus.io v0.24.0 // indirect
	golang.org/x/net v0.5.0 // indirect
	golang.org/x/oauth2 v0.4.0 // indirect
	golang.org/x/sys v0.4.0 // indirect
	golang.org/x/text v0.6.0 // indirect
	google.golang.org/api v0.108.0 // indirect
	google.golang.org/appengine v1.6.7 // indirect
	google.golang.org/genproto v0.0.0-20230125152338-dcaf20b6aeaa // indirect
	google.golang.org/grpc v1.52.1 // indirect
	google.golang.org/protobuf v1.28.1 // indirect
)
```

**Comment by quartzmo** [comment]
Can you try creating a clean registry and using the commented-out Device create loop in my demo, then run your script against the new registry and post here the output?

**Comment by quartzmo** [comment]
@rajasd27 Ping. Any progress getting answers about this unusual behavior?

**Comment by rajasd27** [comment]
Hi Chris - Sorry for the delayed response. I created devices from your code snippet and tried to fetch them and got the same number of devices. We identified the issue and it was related to the GCP devices page UI. Since we had more than 5k devices, we were pressing the next button on the devices page pretty quickly to get the final count and for some reason, the UI shows the number of devices to be greater than the actual count. We thought this was a backend issue but looks like the UI is showing an incorrect count and that's what misled us. 

I think we can go ahead and close this issue since it's no longer related to the SDK. Thank you for all your help & support.
